src/Processors/Executors/PullingPipelineExecutor.cpp

```diff
@@ -0,0 +1,186 @@
+#include <Processors/Executors/PullingPipelineExecutor.h>
+#include <Processors/Executors/PipelineExecutor.h>
+#include <Processors/Formats/LazyOutputFormat.h>
+#include <Processors/Transforms/AggregatingTransform.h>
+#include <Processors/QueryPipeline.h>
+
+#include <Common/setThreadName.h>
+#include <ext/scope_guard.h>
+
+namespace DB
+{
+
+struct PullingPipelineExecutor::Data
+{
+    PipelineExecutorPtr executor;
+    std::exception_ptr exception;
+    std::atomic_bool is_executed = false;
+    std::atomic_bool has_exception = false;
+    ThreadFromGlobalPool thread;
+
+    ~Data()
+    {
+        if (thread.joinable())
+            thread.join();
+    }
+};
+
+PullingPipelineExecutor::PullingPipelineExecutor(QueryPipeline & pipeline_) : pipeline(pipeline_)
+{
+    lazy_format = std::make_shared<LazyOutputFormat>(pipeline.getHeader());
+    pipeline.setOutput(lazy_format);
+}
+
+PullingPipelineExecutor::~PullingPipelineExecutor()
+{
+    try
+    {
+        /// Cancel execution if it wasn't finished.
+        if (data && !data->is_executed)
+            cancel();
+
+        /// Finish lazy format. Otherwise thread.join() may hung.
+        if (!lazy_format->isFinished())
+            lazy_format->finish();
+
+        /// Join thread here to wait for possible exception.
+        if (data && data->thread.joinable())
+            data->thread.join();
+    }
+    catch (...)
+    {
+        tryLogCurrentException("PullingPipelineExecutor");
+    }
+
+    if (data && data->has_exception)
+        tryLogException(std::move(data->exception), "PullingPipelineExecutor");
+}
+
+static void threadFunction(PullingPipelineExecutor::Data & data, ThreadGroupStatusPtr thread_group, size_t num_threads)
+{
+    if (thread_group)
+        CurrentThread::attachTo(thread_group);
+
+    SCOPE_EXIT(
+        if (thread_group)
+            CurrentThread::detachQueryIfNotDetached();
+    );
+
+    setThreadName("QueryPipelineEx");
+
+    try
+    {
+        data.executor->execute(num_threads);
+    }
+    catch (...)
+    {
+        data.exception = std::current_exception();
+        data.has_exception = true;
+    }
+
+    data.is_executed = true;
+}
+
+
+bool PullingPipelineExecutor::pull(Chunk & chunk, uint64_t milliseconds)
+{
+    if (!data)
+    {
+        data = std::make_unique<Data>();
+        data->executor = pipeline.execute();
+
+        auto func = [&, thread_group = CurrentThread::getGroup()]()
+        {
+            threadFunction(*data, thread_group, pipeline.getNumThreads());
+        };
+
+        data->thread = ThreadFromGlobalPool(std::move(func));
+    }
+
+    if (data->has_exception)
+    {
+        /// Finish lazy format in case of exception. Otherwise thread.join() may hung.
+        lazy_format->finish();
+        data->has_exception = false;
+        std::rethrow_exception(std::move(data->exception));
+    }
+
+    if (lazy_format->isFinished())
+        return false;
+
+    chunk = lazy_format->getChunk(milliseconds);
```

Comment:
I don't see any synchronization here. Is it Ok, that there could be a sequence of empty chunks and then normal chunk with data?

Reply:
> I don't see any synchronization here.
It's inside 'lazy_format', in `ConcurrentBoundedQueue`

> Is it Ok, that there could be a sequence of empty chunks and then normal chunk with data?
Yes, because we return special flag, which specifies if reading finished or not.